File: simple-examples/clj/src/com/rpl/agent/langchain4j_agent.clj
(ns com.rpl.agent.langchain4j-agent
  "Demonstrates LangChain4j chat model integration with agent-o-rama.

  Features demonstrated:
  - OpenAI chat model configuration and usage
  - Message handling with UserMessage, AiMessage, SystemMessage
  - Chat request customization with temperature and tokens
  - Conversation state management across nodes"
  (:require
   [com.rpl.agent-o-rama :as aor]
   [com.rpl.agent-o-rama.langchain4j :as lc4j]
   [com.rpl.rama :as rama]
   [com.rpl.rama.test :as rtest])
  (:import
   [dev.langchain4j.data.message
    AiMessage
    SystemMessage
    UserMessage]
   [dev.langchain4j.model.openai
    OpenAiChatModel]))

;;; Agent module demonstrating LangChain4j integration
(aor/defagentmodule LangChain4jAgentModule
  [topology]

  ;; Declare OpenAI API key as agent object
  (aor/declare-agent-object
   topology
   "openai-api-key"
   (or (System/getenv "OPENAI_API_KEY") "demo-key"))

  ;; Build OpenAI chat model with configuration
  (aor/declare-agent-object-builder
   topology
   "openai-model"
   (fn [setup]
     (-> (OpenAiChatModel/builder)
         (.apiKey (aor/get-agent-object setup "openai-api-key"))
         (.modelName "gpt-4o-mini")
         (.temperature 0.7)
         (.maxTokens 500)
         .build)))

  (->
    topology
    (aor/new-agent "LangChain4jAgent")

    ;; Initial chat node that starts conversation
    (aor/node
     "start-chat"
     "continue-chat"
     (fn [agent-node {:keys [system-prompt user-message]}]
       (println "Starting chat conversation...")

       (let [model         (aor/get-agent-object agent-node "openai-model")
             messages      [(SystemMessage. system-prompt)
                            (UserMessage. user-message)]

             ;; Send chat request to OpenAI
             response      (lc4j/chat model
                                      (lc4j/chat-request messages
                                                         {:temperature 0.7
                                                          :max-output-tokens
                                                          200}))
             ai-message    (.aiMessage response)
             response-text (.text ai-message)]

         (println "AI Response:" response-text)

         ;; Continue conversation with context
         (aor/emit! agent-node
                    "continue-chat"
                    {:conversation-history (conj messages ai-message)
                     :last-response        response-text
                     :turn-count           1}))))

    ;; Continue conversation with follow-up
    (aor/node
     "continue-chat"
     "analyze-conversation"
     (fn [agent-node {:keys [conversation-history last-response turn-count]}]
       (println (format "Continuing conversation (turn %d)..."
                        (inc turn-count)))

       (let [model             (aor/get-agent-object agent-node "openai-model")
             follow-up-message (UserMessage.
                                "Can you provide more details or an example?")
             updated-history   (conj conversation-history follow-up-message)

             ;; Send follow-up with full conversation history
             response          (lc4j/chat model
                                          (lc4j/chat-request updated-history
                                                             {:temperature 0.5
                                                              :max-output-tokens
                                                              300}))
             ai-message        (.aiMessage response)
             response-text     (.text ai-message)]

         (println "Follow-up Response:" response-text)

         (aor/emit! agent-node
                    "analyze-conversation"
                    {:full-conversation  (conj updated-history ai-message)
                     :initial-response   last-response
                     :follow-up-response response-text
                     :total-turns        (inc turn-count)}))))

    ;; Analyze and summarize conversation
    (aor/node
     "analyze-conversation"
     nil
     (fn [agent-node
          {:keys [full-conversation initial-response
                  follow-up-response total-turns]}]
       (println "Analyzing conversation...")

       (let [model           (aor/get-agent-object agent-node "openai-model")
             ;; Create analysis prompt
             analysis-prompt (str
                              "Summarize this conversation in one sentence: "
                              initial-response
                              " ... " follow-up-response)

             ;; Get summary
             summary-response (lc4j/chat model
                                         (lc4j/chat-request
                                          [(UserMessage. analysis-prompt)]
                                          {:temperature       0.3
                                           :max-output-tokens 50}))
             summary         (.text (.aiMessage summary-response))]

         (println "Conversation Summary:" summary)

         (aor/result! agent-node
                      {:action               "chat-complete"
                       :total-turns          (inc total-turns)
                       :message-count        (count full-conversation)
                       :initial-response     initial-response
                       :follow-up-response   follow-up-response
                       :conversation-summary summary
                       :model-used           "gpt-4o-mini"
                       :processed-at         (System/currentTimeMillis)}))))))

(defn -main
  "Run the LangChain4j agent example"
  [& _args]
  (if (System/getenv "OPENAI_API_KEY")
    (with-open [ipc (rtest/create-ipc)]
      (rtest/launch-module! ipc LangChain4jAgentModule {:tasks 1 :threads 1})

      (let [manager (aor/agent-manager ipc
                                       (rama/get-module-name
                                        LangChain4jAgentModule))
            agent   (aor/agent-client manager "LangChain4jAgent")]

        (println "LangChain4j Agent Example:")
        (println "Demonstrating multi-turn conversation with OpenAI\n")

        (let
          [result
           (aor/agent-invoke
            agent
            {:system-prompt
             "You are a helpful assistant that explains technical concepts clearly and concisely."
             :user-message "What is functional programming?"})]

          (println "\n=== Final Results ===")
          (println "  Action:" (:action result))
          (println "  Total turns:" (:total-turns result))
          (println "  Total messages:" (:message-count result))
          (println "  Model used:" (:model-used result))
          (println "\n  Initial response snippet:"
                   (subs (:initial-response result)
                         0
                         (min 100 (count (:initial-response result))))
                   "...")
          (println "\n  Follow-up response snippet:"
                   (subs (:follow-up-response result)
                         0
                         (min 100 (count (:follow-up-response result))))
                   "...")
          (println "\n  Summary:" (:conversation-summary result)))

        (println "\nNotice how:")
        (println
         "- OpenAI model is configured with temperature and token limits")
        (println "- Conversation history is maintained across nodes")
        (println "- Different parameters can be used for each request")
        (println "- Full LangChain4j message types are supported")))

    (do
      (println "LangChain4j Agent Example:")
      (println "OPENAI_API_KEY environment variable not set.")
      (println "Please set your OpenAI API key to run this example:")
      (println "  export OPENAI_API_KEY=your-api-key-here"))))

File: examples/java/basic/src/main/java/com/rpl/agent/basic/LangChain4jAgent.java
package com.rpl.agent.basic;

import com.rpl.agentorama.AgentClient;
import com.rpl.agentorama.AgentManager;
import com.rpl.agentorama.AgentNode;
import com.rpl.agentorama.AgentsModule;
import com.rpl.agentorama.AgentsTopology;
import com.rpl.agentorama.langchain4j.LangChain4j;
import com.rpl.agentorama.ops.RamaVoidFunction2;
import com.rpl.rama.test.InProcessCluster;
import com.rpl.rama.test.LaunchConfig;
import dev.langchain4j.data.message.SystemMessage;
import dev.langchain4j.data.message.UserMessage;
import dev.langchain4j.model.openai.OpenAiChatModel;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Java example demonstrating LangChain4j chat model integration with agent-o-rama.
 *
 * <p>Features demonstrated:
 *
 * <ul>
 *   <li>OpenAI chat model configuration as agent object
 *   <li>Message handling with SystemMessage and UserMessage
 *   <li>Chat request with temperature and token limits
 *   <li>Simple single-node chat completion
 * </ul>
 *
 * <p>All required classes are defined as nested classes within this single file for simplicity and
 * self-containment.
 */
public class LangChain4jAgent {

  /** Agent Module demonstrating LangChain4j integration. */
  public static class LangChain4jModule extends AgentsModule {

    @Override
    protected void defineAgents(AgentsTopology topology) {
      // Declare OpenAI API key as agent object
      topology.declareAgentObject("openai-api-key", System.getenv("OPENAI_API_KEY"));

      // Build OpenAI chat model with configuration
      topology.declareAgentObjectBuilder(
          "openai-model",
          setup -> {
            String apiKey = (String) setup.getAgentObject("openai-api-key");
            return OpenAiChatModel.builder()
                .apiKey(apiKey)
                .modelName("gpt-4o-mini")
                .temperature(0.7)
                .maxTokens(500)
                .build();
          });

      topology.newAgent("LangChain4jAgent").node("chat", null, new ChatFunction());
    }
  }

  /** Node function that sends user message to OpenAI and returns response. */
  public static class ChatFunction implements RamaVoidFunction2<AgentNode, String> {

    @Override
    public void invoke(AgentNode agentNode, String userMessage) {
      OpenAiChatModel model = (OpenAiChatModel) agentNode.getAgentObject("openai-model");

      // Prepare messages
      SystemMessage systemMessage = new SystemMessage("You are a helpful assistant.");
      UserMessage userMsg = new UserMessage(userMessage);

      // Create chat request options
      Map<String, Object> options = new HashMap<>();
      options.put("temperature", 0.7);
      options.put("max-output-tokens", 200);

      // Send chat request to OpenAI
      var response =
          LangChain4j.chat(
              model, LangChain4j.chatRequest(Arrays.asList(systemMessage, userMsg), options));
      String responseText = response.aiMessage().text();

      agentNode.result(responseText);
    }
  }

  public static void main(String[] args) throws Exception {
    String apiKey = System.getenv("OPENAI_API_KEY");
    if (apiKey == null || apiKey.trim().isEmpty()) {
      System.out.println("LangChain4j Agent Example:");
      System.out.println("OPENAI_API_KEY environment variable not set.");
      System.out.println("Please set your OpenAI API key to run this example:");
      System.out.println("  export OPENAI_API_KEY=your-api-key-here");
      return;
    }

    System.out.println("Starting LangChain4j Agent Example...");

    try (InProcessCluster ipc = InProcessCluster.create()) {
      // Launch the agent module
      LangChain4jModule module = new LangChain4jModule();
      ipc.launchModule(module, new LaunchConfig(1, 1));

      // Get agent manager and client
      String moduleName = module.getModuleName();
      AgentManager manager = AgentManager.create(ipc, moduleName);
      AgentClient agent = manager.getAgentClient("LangChain4jAgent");

      System.out.println("LangChain4j Agent Example:");
      System.out.println("Sending message to OpenAI...\n");

      String result = (String) agent.invoke("What is agent-o-rama?");
      System.out.println("User: What is agent-o-rama?");
      System.out.println("\nAssistant: " + result);

      System.out.println("\nNotice how:");
      System.out.println("- OpenAI model is configured as an agent object");
      System.out.println("- Single node handles the complete chat interaction");
      System.out.println("- Temperature and token limits are customizable");
    }
  }
}
